src/main/java/com/salesforce/dynamodbv2/mt/util/CachingAmazonDynamoDbStreams.java

```diff
@@ -0,0 +1,964 @@
+     */
+    private static final class IteratorPosition implements Comparable<IteratorPosition>, Predicate<Record> {
+
+        private final String streamArn;
+        private final String shardId;
+        private final BigInteger sequenceNumber;
+
+        IteratorPosition(String streamArn, String shardId, BigInteger sequenceNumber) {
+            this.streamArn = checkNotNull(streamArn);
+            this.shardId = checkNotNull(shardId);
+            this.sequenceNumber = checkNotNull(sequenceNumber);
+        }
+
+        /**
+         * Implementation of Predicate interface so this position can be used directly to filter records. Also enables
+         * negating the precedes check if this position succeeds records.
+         */
+        @Override
+        public boolean test(Record record) {
+            return precedes(record);
+        }
+
+        /**
+         * Checks if this iterator is before or at the given record position in the shard. Note: the client is
+         * responsible for making sure the record is from the same stream and shard; records do not carry shard
+         * information, so the implementation cannot verify they match. The behavior of comparing iterator positions and
+         * records from different shards in unspecified.
+         *
+         * @param record Record to check.
+         * @return true if the sequence number of the record is greater or equal to the position sequence number.
+         */
+        boolean precedes(Record record) {
+            return sequenceNumber.compareTo(parseSequenceNumber(record)) <= 0;
+        }
+
+        /**
+         * Returns true if this position precedes any records in the given result. Since DynamoDB orders records by
+         * sequence number that equivalent to checking whether this position precedes the last record.
+         *
+         * @param result Result to check. Records list must not be empty.
+         * @return True if this position precedes any records in the given result, false otherwise.
+         */
+        boolean precedesAny(GetRecordsResult result) {
+            assert !result.getRecords().isEmpty();
+            return precedes(getLast(result.getRecords()));
+        }
+
+        /**
+         * Returns the position immediately following this one, i.e., the next sequence number in the same stream
+         * shard.
+         *
+         * @return Next iterator position.
+         */
+        IteratorPosition next() {
+            return new IteratorPosition(streamArn, shardId, sequenceNumber.add(ONE));
+        }
+
+        /**
+         * Returns the position immediately following the last record in the given result.
+         *
+         * @param result Result to position after. Record list must not be empty.
+         * @return IteratorPosition immediately following the last record sequence number in the result.
+         */
+        IteratorPosition next(GetRecordsResult result) {
```

Comment:
Maybe rename this or do something to distinguish from next() which is strictly dependent on the sequence.  Same comment on nextShardIterator().